File: examples/agent-starter.ts
/**
 * Conductor Agent - TypeScript Starter
 *
 * This is a sample agent that polls the Conductor API for tasks and executes them.
 * Customize this script based on your agent's capabilities.
 *
 * Usage:
 *     ts-node agent-starter.ts --agent-id YOUR_AGENT_ID --api-key YOUR_API_KEY
 *
 * Requirements:
 *     npm install typescript ts-node @types/node
 *     npm install @anthropic-ai/sdk openai
 */

import { Task, TaskLog, CompleteTaskRequest, FailTaskRequest, PollTaskRequest, AgentHeartbeatRequest } from '../types'

interface AgentConfig {
  agentId: string
  apiKey: string
  capabilities: string[]
  baseUrl?: string
  pollInterval?: number
}

interface TaskOutput {
  [key: string]: any
}

class ConductorAgent {
  private agentId: string
  private apiKey: string
  private capabilities: string[]
  private baseUrl: string
  private pollInterval: number
  private running: boolean = true

  constructor(config: AgentConfig) {
    this.agentId = config.agentId
    this.apiKey = config.apiKey
    this.capabilities = config.capabilities
    this.baseUrl = config.baseUrl || 'http://localhost:3000'
    this.pollInterval = config.pollInterval || 5000 // 5 seconds
  }

  /**
   * Poll the API for an available task
   */
  async pollForTask(): Promise<Task | null> {
    try {
      const response = await fetch(`${this.baseUrl}/api/tasks/poll`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          agent_id: this.agentId,
          capabilities: this.capabilities,
        } as PollTaskRequest),
      })

      if (response.ok) {
        const data = await response.json()
        if (data.success && data.data) {
          return data.data as Task
        }
      }
      return null
    } catch (error) {
      console.error('Error polling for task:', error)
      return null
    }
  }

  /**
   * Send heartbeat to indicate agent is alive
   */
  async sendHeartbeat(status: 'idle' | 'active' | 'busy' | 'offline' | 'error' = 'active'): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/api/agents/heartbeat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          agent_id: this.agentId,
          status,
        } as AgentHeartbeatRequest),
      })
      return response.ok
    } catch (error) {
      console.error('Error sending heartbeat:', error)
      return false
    }
  }

  /**
   * Log a message for a task
   */
  async logTaskMessage(
    taskId: string,
    message: string,
    level: 'info' | 'warning' | 'error' | 'debug' = 'info',
    data: Record<string, any> = {}
  ): Promise<void> {
    try {
      await fetch(`${this.baseUrl}/api/tasks/${taskId}/logs`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          agent_id: this.agentId,
          level,
          message,
          data,
        }),
      })
    } catch (error) {
      console.error('Error logging task message:', error)
    }
  }

  /**
   * Mark task as completed
   */
  async completeTask(taskId: string, outputData: TaskOutput): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/api/tasks/${taskId}/complete`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          agent_id: this.agentId,
          output_data: outputData,
        } as CompleteTaskRequest),
      })
      return response.ok
    } catch (error) {
      console.error('Error completing task:', error)
      return false
    }
  }

  /**
   * Mark task as failed
   */
  async failTask(taskId: string, errorMessage: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/api/tasks/${taskId}/fail`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          agent_id: this.agentId,
          error_message: errorMessage,
        } as FailTaskRequest),
      })
      return response.ok
    } catch (error) {
      console.error('Error failing task:', error)
      return false
    }
  }

  /**
   * Execute a task. Override this method in your implementation.
   */
  async executeTask(task: Task): Promise<TaskOutput> {
    throw new Error('executeTask() must be implemented by subclass')
  }

  /**
   * Main agent loop
   */
  async run(): Promise<void> {
    console.log(`Agent ${this.agentId} starting...`)
    console.log(`Capabilities: ${this.capabilities.join(', ')}`)

    let heartbeatCounter = 0

    while (this.running) {
      try {
        // Send heartbeat every 6 polls (30 seconds if poll interval is 5s)
        if (heartbeatCounter % 6 === 0) {
          await this.sendHeartbeat('active')
          console.debug('Sent heartbeat')
        }
        heartbeatCounter++

        // Poll for task
        const task = await this.pollForTask()

        if (task) {
          console.log(`Received task: ${task.title}`)

          await this.logTaskMessage(task.id, `Task assigned to agent ${this.agentId}`)

          try {
            // Update heartbeat to busy
            await this.sendHeartbeat('busy')

            // Execute the task
            await this.logTaskMessage(task.id, 'Starting task execution')
            const outputData = await this.executeTask(task)

            // Complete the task
            await this.logTaskMessage(task.id, 'Task completed successfully')
            await this.completeTask(task.id, outputData)
            console.log(`Task ${task.id} completed successfully`)
          } catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error)
            console.error(`Task execution failed: ${errorMessage}`)
            await this.logTaskMessage(task.id, `Task execution failed: ${errorMessage}`, 'error')
            await this.failTask(task.id, errorMessage)
          } finally {
            // Reset to active
            await this.sendHeartbeat('active')
          }
        } else {
          console.debug('No tasks available')
        }

        // Wait before next poll
        await new Promise((resolve) => setTimeout(resolve, this.pollInterval))
      } catch (error) {
        console.error('Error in agent loop:', error)
        await new Promise((resolve) => setTimeout(resolve, this.pollInterval))
      }
    }
  }

  /**
   * Gracefully shutdown the agent
   */
  async shutdown(): Promise<void> {
    console.log('Shutting down agent...')
    this.running = false
    await this.sendHeartbeat('offline')
  }
}

/**
 * Example agent implementation
 */
class ExampleAgent extends ConductorAgent {
  constructor(agentId: string, apiKey: string, baseUrl?: string) {
    super({
      agentId,
      apiKey,
      capabilities: ['coding', 'analysis', 'documentation'],
      baseUrl,
    })

    // Initialize your AI client here (Anthropic, OpenAI, etc.)
    // this.aiClient = new Anthropic({ apiKey })
  }

  async executeTask(task: Task): Promise<TaskOutput> {
    const { id: taskId, title, description, input_data, type } = task

    console.log(`Executing ${type} task: ${title}`)

    // Log progress
    await this.logTaskMessage(taskId, 'Analyzing task requirements')

    // Here you would call your AI model
    // For example with Anthropic:
    // const response = await this.aiClient.messages.create({
    //   model: 'claude-sonnet-4',
    //   messages: [{
    //     role: 'user',
    //     content: `Task: ${title}\nDescription: ${description}\nInput: ${JSON.stringify(input_data)}`
    //   }]
    // })

    // Simulate work
    await new Promise((resolve) => setTimeout(resolve, 2000))
    await this.logTaskMessage(taskId, 'Processing task with AI model')

    await new Promise((resolve) => setTimeout(resolve, 2000))
    await this.logTaskMessage(taskId, 'Generating output')

    await new Promise((resolve) => setTimeout(resolve, 1000))

    // Return output data
    return {
      result: `Completed ${type} task: ${title}`,
      task_type: type,
      execution_time: '5 seconds',
      model: 'example-model',
      success: true,
    }
  }
}

/**
 * Main entry point
 */
async function main() {
  const args = process.argv.slice(2)
  const agentId = args.find((arg) => arg.startsWith('--agent-id='))?.split('=')[1]
  const apiKey = args.find((arg) => arg.startsWith('--api-key='))?.split('=')[1]
  const baseUrl = args.find((arg) => arg.startsWith('--base-url='))?.split('=')[1] || 'http://localhost:3000'

  if (!agentId || !apiKey) {
    console.error('Usage: ts-node agent-starter.ts --agent-id=AGENT_ID --api-key=API_KEY [--base-url=BASE_URL]')
    process.exit(1)
  }

  // Create agent
  const agent = new ExampleAgent(agentId, apiKey, baseUrl)

  // Handle graceful shutdown
  process.on('SIGINT', async () => {
    await agent.shutdown()
    process.exit(0)
  })

  process.on('SIGTERM', async () => {
    await agent.shutdown()
    process.exit(0)
  })

  // Run agent
  await agent.run()
}

// Run if this is the main module
if (require.main === module) {
  main().catch((error) => {
    console.error('Fatal error:', error)
    process.exit(1)
  })
}

export { ConductorAgent, ExampleAgent }

File: examples/agent-starter.py
#!/usr/bin/env python3
"""
Conductor Agent - Python Starter

This is a sample agent that polls the Conductor API for tasks and executes them.
Customize this script based on your agent's capabilities.

Usage:
    python agent-starter.py --agent-id YOUR_AGENT_ID --api-key YOUR_API_KEY

Requirements:
    pip install requests anthropic openai
"""

import argparse
import json
import logging
import time
from typing import Dict, Any, Optional
import requests

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

class ConductorAgent:
    """Base class for Conductor agents"""

    def __init__(
        self,
        agent_id: str,
        api_key: str,
        capabilities: list[str],
        base_url: str = "http://localhost:3000",
        poll_interval: int = 5
    ):
        self.agent_id = agent_id
        self.api_key = api_key
        self.capabilities = capabilities
        self.base_url = base_url
        self.poll_interval = poll_interval
        self.running = True

    def poll_for_task(self) -> Optional[Dict[str, Any]]:
        """Poll the API for an available task"""
        try:
            response = requests.post(
                f"{self.base_url}/api/tasks/poll",
                json={
                    "agent_id": self.agent_id,
                    "capabilities": self.capabilities
                },
                headers={"Content-Type": "application/json"}
            )

            if response.status_code == 200:
                data = response.json()
                if data.get("success") and data.get("data"):
                    return data["data"]
            return None

        except Exception as e:
            logger.error(f"Error polling for task: {e}")
            return None

    def send_heartbeat(self, status: str = "active") -> bool:
        """Send heartbeat to indicate agent is alive"""
        try:
            response = requests.post(
                f"{self.base_url}/api/agents/heartbeat",
                json={
                    "agent_id": self.agent_id,
                    "status": status
                },
                headers={"Content-Type": "application/json"}
            )
            return response.status_code == 200
        except Exception as e:
            logger.error(f"Error sending heartbeat: {e}")
            return False

    def log_task_message(self, task_id: str, message: str, level: str = "info", data: Optional[Dict] = None):
        """Log a message for a task"""
        try:
            requests.post(
                f"{self.base_url}/api/tasks/{task_id}/logs",
                json={
                    "agent_id": self.agent_id,
                    "level": level,
                    "message": message,
                    "data": data or {}
                },
                headers={"Content-Type": "application/json"}
            )
        except Exception as e:
            logger.error(f"Error logging task message: {e}")

    def complete_task(self, task_id: str, output_data: Dict[str, Any]) -> bool:
        """Mark task as completed"""
        try:
            response = requests.post(
                f"{self.base_url}/api/tasks/{task_id}/complete",
                json={
                    "agent_id": self.agent_id,
                    "output_data": output_data
                },
                headers={"Content-Type": "application/json"}
            )
            return response.status_code == 200
        except Exception as e:
            logger.error(f"Error completing task: {e}")
            return False

    def fail_task(self, task_id: str, error_message: str) -> bool:
        """Mark task as failed"""
        try:
            response = requests.post(
                f"{self.base_url}/api/tasks/{task_id}/fail",
                json={
                    "agent_id": self.agent_id,
                    "error_message": error_message
                },
                headers={"Content-Type": "application/json"}
            )
            return response.status_code == 200
        except Exception as e:
            logger.error(f"Error failing task: {e}")
            return False

    def execute_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """
        Execute the given task. Override this method in your agent implementation.

        Args:
            task: The task object containing title, description, input_data, etc.

        Returns:
            Dictionary containing the output data
        """
        raise NotImplementedError("Subclasses must implement execute_task()")

    def run(self):
        """Main agent loop"""
        logger.info(f"Agent {self.agent_id} starting...")
        logger.info(f"Capabilities: {', '.join(self.capabilities)}")

        heartbeat_counter = 0

        while self.running:
            try:
                # Send heartbeat every 6 polls (30 seconds if poll_interval is 5)
                if heartbeat_counter % 6 == 0:
                    self.send_heartbeat("active")
                    logger.debug("Sent heartbeat")
                heartbeat_counter += 1

                # Poll for task
                task = self.poll_for_task()

                if task:
                    task_id = task["id"]
                    logger.info(f"Received task: {task['title']}")

                    self.log_task_message(task_id, f"Task assigned to agent {self.agent_id}")

                    try:
                        # Update heartbeat to busy
                        self.send_heartbeat("busy")

                        # Execute the task
                        self.log_task_message(task_id, "Starting task execution")
                        output_data = self.execute_task(task)

                        # Complete the task
                        self.log_task_message(task_id, "Task completed successfully")
                        self.complete_task(task_id, output_data)
                        logger.info(f"Task {task_id} completed successfully")

                    except Exception as e:
                        logger.error(f"Task execution failed: {e}")
                        self.log_task_message(task_id, f"Task execution failed: {str(e)}", level="error")
                        self.fail_task(task_id, str(e))

                    finally:
                        # Reset to active
                        self.send_heartbeat("active")
                else:
                    logger.debug("No tasks available")

                # Wait before next poll
                time.sleep(self.poll_interval)

            except KeyboardInterrupt:
                logger.info("Shutting down agent...")
                self.running = False
                self.send_heartbeat("offline")
                break
            except Exception as e:
                logger.error(f"Error in agent loop: {e}")
                time.sleep(self.poll_interval)


class ExampleAgent(ConductorAgent):
    """Example agent implementation"""

    def __init__(self, agent_id: str, api_key: str, **kwargs):
        super().__init__(
            agent_id=agent_id,
            api_key=api_key,
            capabilities=["coding", "analysis", "documentation"],
            **kwargs
        )
        # Initialize your AI client here (Anthropic, OpenAI, etc.)
        # self.ai_client = anthropic.Anthropic(api_key=api_key)

    def execute_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a task using AI"""
        task_id = task["id"]
        title = task["title"]
        description = task.get("description", "")
        input_data = task.get("input_data", {})
        task_type = task.get("type", "feature")

        logger.info(f"Executing {task_type} task: {title}")

        # Log progress
        self.log_task_message(task_id, f"Analyzing task requirements")

        # Here you would call your AI model
        # For example with Anthropic:
        # response = self.ai_client.messages.create(
        #     model="claude-sonnet-4",
        #     messages=[{
        #         "role": "user",
        #         "content": f"Task: {title}\nDescription: {description}\nInput: {json.dumps(input_data)}"
        #     }]
        # )

        # Simulate work
        time.sleep(2)
        self.log_task_message(task_id, "Processing task with AI model")
        time.sleep(2)
        self.log_task_message(task_id, "Generating output")
        time.sleep(1)

        # Return output data
        return {
            "result": f"Completed {task_type} task: {title}",
            "task_type": task_type,
            "execution_time": "5 seconds",
            "model": "example-model",
            "success": True
        }


def main():
    parser = argparse.ArgumentParser(description="Run a Conductor agent")
    parser.add_argument("--agent-id", required=True, help="Agent ID from Conductor")
    parser.add_argument("--api-key", required=True, help="AI API key (Anthropic/OpenAI)")
    parser.add_argument("--base-url", default="http://localhost:3000", help="Conductor API base URL")
    parser.add_argument("--poll-interval", type=int, default=5, help="Polling interval in seconds")

    args = parser.parse_args()

    # Create and run agent
    agent = ExampleAgent(
        agent_id=args.agent_id,
        api_key=args.api_key,
        base_url=args.base_url,
        poll_interval=args.poll_interval
    )

    agent.run()


if __name__ == "__main__":
    main()
